### ISSUE TITLE
eCLM-PDAF on Stages/2024

### ISSUE THREAD

**Issue by jjokella** [HPSCTerrSys/TSMP2#36]
Background: On `Stages/2023` and for eCLM-PDAF,  Buffer-Overflow-Errors are now very frequent on both JURECA und JUWELS.

The simplest solution seems to be to switch to `Stages/2024`, i.e. compile with branch `stages-2024`. 

This can be done by:
```
git clone --branch stages-2024 https://github.com/HPSCTerrSys/TSMP2.git
cd TSMP2
./build_tsmp2.sh --eCLM --PDAF --build_type DEBUG
```

However, this leads to errors in the linking stage:

```
mpif90  -O0 -g -traceback -xHost -r8  -o /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/bin/tsmp-pdaf \
mod_tsmp.o mod_parallel_pdaf.o mod_assimilation.o parser_mpi.o mod_read_obs.o pdaf_terrsysmp.o integrate_pdaf.o init_parallel_pdaf.o finalize_pdaf.o init_pdaf.o init_pdaf_parse.o init_pdaf_info.o assimilate_pdaf.o init_ens.o prepoststep_ens_pdaf.o next_observation_pdaf.o distribute_state_pdaf.o collect_state_pdaf.o init_obs_pdaf.o init_dim_obs_pdaf.o obs_op_pdaf.o prodrinva_pdaf.o init_obsvar_pdaf.o add_obs_error_pdaf.o init_obscovar_pdaf.o  init_n_domains_pdaf.o init_dim_l_pdaf.o g2l_state_pdaf.o l2g_state_pdaf.o prodrinva_l_pdaf.o init_obs_l_pdaf.o init_dim_obs_l_pdaf.o g2l_obs_pdaf.o init_obsvar_l_pdaf.o init_dim_obs_f_pdaf.o init_obs_f_pdaf.o obs_op_f_pdaf.o localize_covar_pdaf.o \
-L/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/pdaf/lib -lpdaf-d -Wl,--start-group    -qmkl -Wl,--end-group -Wl,-rpath -Wl,/p/software/juwels/stages/2024/software/psmpi/5.9.2-1-intel-compilers-2023.2.1/lib -Wl,--enable-new-dtags -Wl,-rpath-link=/p/software/juwels/stages/2024/software/pscom/5-default-GCCcore-12.3.0/lib -L/p/software/juwels/stages/2024/software/psmpi/5.9.2-1-intel-compilers-2023.2.1/lib -lmpi -ldl -lpthread -lcudart -lstdc++ -lcuda -lrt -lgpfs -lpmix -lmpich -fiopenmp -L/p/software/juwels/stages/2024/software/netCDF-Fortran/4.6.1-ipsmpi-2023a/lib -lnetcdff -L/p/software/juwels/stages/2024/software/netCDF-Fortran/4.6.1-ipsmpi-2023a/lib -lnetcdff -lnetcdf -L/p/software/juwels/stages/2024/software/netCDF/4.9.2-ipsmpi-2023a/lib64 -lnetcdf -lhdf5_hl -lhdf5 -lm -lz -lblosc -lzstd -lbz2 -lsz -lcurl -lpnetcdf -lxml2 /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib/libmodel.a -L/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib -L/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib -leclm -lclm -ldatm -lglc -lgptl -lice -lmosart -locn -lpio -lwav -lesp -lmct2 -lmpeu -lcsm_share
ld: warning: size of symbol `outdir' changed from 1 in /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib/libmodel.a(enkf_clm_mod_5.o) to 100 in /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib/libmodel.a(wrapper_tsmp.o)
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib/libeclm.a(cime_comp_mod.F90.o): in function `cime_comp_mod_mp_cime_pre_init1_':
/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/eclm/cime_comp_mod.F90:587: undefined reference to `__msan_param_tls'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/eclm/cime_comp_mod.F90:587: undefined reference to `__msan_param_origin_tls'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/eclm/cime_comp_mod.F90:587: undefined reference to `__msan_set_alloca_origin_with_descr'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/eclm/cime_comp_mod.F90:587: undefined reference to `__msan_set_alloca_origin_with_descr'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/eclm/cime_comp_mod.F90:587: undefined reference to `__msan_set_alloca_origin_with_descr'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/eclm/cime_comp_mod.F90:587: undefined reference to `__msan_set_alloca_origin_with_descr'

.
.
.

ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/csm_share/mct/shr_expr_parser_mod.F90:180: undefined reference to `__msan_va_arg_overflow_size_tls'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/csm_share/mct/shr_expr_parser_mod.F90:180: undefined reference to `__msan_warning_with_origin_noreturn'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/csm_share/mct/shr_expr_parser_mod.F90:180: undefined reference to `__msan_warning_with_origin_noreturn'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/csm_share/mct/shr_expr_parser_mod.F90:176: undefined reference to `__msan_warning_with_origin_noreturn'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/csm_share/mct/shr_expr_parser_mod.F90:176: undefined reference to `__msan_warning_with_origin_noreturn'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/csm_share/mct/shr_expr_parser_mod.F90:176: undefined reference to `__msan_warning_with_origin_noreturn'
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib/libcsm_share.a(shr_expr_parser_mod.F90.o):/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/eCLM/src/csm_share/mct/shr_expr_parser_mod.F90:179: more undefined references to `__msan_warning_with_origin_noreturn' follow
ld: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/lib/libcsm_share.a(shr_expr_parser_mod.F90.o): in function `msan.module_ctor':
ifxV01R1W.i90:(.text.msan.module_ctor+0x2): undefined reference to `__msan_init'
make[3]: *** [Makefile:147: /p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bin/JUWELS_eCLM-PDAF/bin/tsmp-pdaf] Error 1
make[3]: Leaving directory '/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/models/pdaf/interface/framework'
gmake[2]: *** [CMakeFiles/PDAF-Framework.dir/build.make:89: PDAF-Framework/src/PDAF-Framework-stamp/PDAF-Framework-build] Error 2
gmake[2]: Leaving directory '/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bld/JUWELS_eCLM-PDAF'
gmake[1]: *** [CMakeFiles/Makefile2:170: CMakeFiles/PDAF-Framework.dir/all] Error 2
gmake[1]: Leaving directory '/p/scratch/cjibg36/keller5/DA/TSMP-PDAF/TSMP2_JUWELS_2024_11_11_stages-2024_eCLM-PDAF_DEBUG/bld/JUWELS_eCLM-PDAF'
gmake: *** [Makefile:94: all] Error 2
```

This could be related: https://community.intel.com/t5/Intel-Fortran-Compiler/Linking-errors-when-using-memory-sanitizer-in-fortran-project/m-p/1521567

- `-check all -check uninit` are candidates that could cause this it seems

**Comment by jjokella** [comment]
The issue may be related to these settings for eCLM

https://github.com/HPSCTerrSys/eCLM/blob/c984adaa550d0cd6e232b8c102274031ec7e83f2/cmake/SetBuildOptions.cmake#L32-L39

**Comment by jjokella** [comment]
Update: The `DEBUG` compilation works, if I set
```cmake
set(CMAKE_Fortran_FLAGS_DEBUG "-O0 -g")
```
for eCLM.

**Comment by jjokella** [comment]
Update: The `DEBUG` compilation works, if I set
```cmake
set(CMAKE_Fortran_FLAGS_DEBUG "-O0 -g")
```
for eCLM.

**Comment by kvrigor** [comment]
> 
> in particular Line 38 for debug settings
> 
> ```cmake
> set(CMAKE_Fortran_FLAGS_DEBUG "-O0 -g -check uninit -check bounds -check pointers -fpe0 -check noarg_temp_created")
> ```

Could you try changing `-check uninit` to `-check all`? The [Intel ifort to ifx Porting Guide](https://www.intel.com/content/www/us/en/developer/articles/guide/porting-guide-for-ifort-to-ifx.html) also suggests this option for guaranteeing correctness.

**Comment by jjokella** [comment]
I think I tracked down the issue from the PDAF side, which is that the same compiler options were given to Fortran/C compilers. I will add a PR that makes this more flexible and then I can make the compilation run through! (Tested so far by locally changing)

**Comment by jjokella** [comment]
I still have a linking error in the debug compilation:
```
ld: /p/software/jurecadc/stages/2024/software/cURL/8.0.1-GCCcore-12.3.0/lib/libcurl.a(libcurl_la-content_encoding.o): undefined reference to symbol 'ZSTD_isError'
ld: /p/software/jurecadc/stages/2024/software/Blosc/1.21.5-GCCcore-12.3.0/lib/libblosc.so.1: error adding symbols: DSO missing from command line
```
It may have something to do with the NetCDF loading, maybe I have to load PNetCDF explicitly?

```
NetCDF_F90_STATIC_LDFLAGS:INTERNAL=-L/p/software/jurecadc/stages/2024/software/netCDF-Fortran/4.6.1-ipsmpi-2023a/lib;-lnetcdff;-L/p/software/jurecadc/stages/2024/software/netCDF-Fortran/4.6.1-ipsmpi-2023a/lib;-lnetcdff;-lnetcdf;-L/p/software/jurecadc/stages/2024/software/netCDF/4.9.2-ipsmpi-2023a/lib64;-lnetcdf;-lhdf5_hl;-lhdf5;-lm;-lz;-lblosc;-lzstd;-lbz2;-lsz;-lcurl;-lpnetcdf;-lxml2
```

**Comment by kvrigor** [comment]
Does the `ld` error occur in eCLM or PDAF? Perhaps we can discuss this on https://github.com/HPSCTerrSys/TSMP2/pull/37 to make it easy to comment on the relevant lines in the source code

**Comment by jjokella** [comment]
This issue is resolved by #37 and #38
